File: src/main/python/AoC2015_16.py
#! /usr/bin/env python3
#
# Advent of Code 2015 Day 16
#

import itertools
import operator
import sys
from collections import defaultdict
from dataclasses import dataclass
from typing import Self
from typing import cast

from aoc.common import InputData
from aoc.common import SolutionBase

EQ, LT, GT = "eq", "lt", "gt"
OPS = {EQ: operator.eq, LT: operator.lt, GT: operator.gt}


@dataclass(frozen=True)
class AuntSue:
    nbr: int
    things: dict[str, int]

    @classmethod
    def from_input(cls, line: str) -> Self:
        splits = line.split(", ")
        nbr = splits[0].split(": ")[0].split(" ")[1]
        splits = line[len("Sue " + nbr + ": ") :].split(", ")
        things = {
            thing: int(amount)
            for thing, amount in (split.split(": ") for split in splits)
        }
        return cls(int(nbr), things)

    def get_thing(self, thing: str) -> int | None:
        return self.things.get(thing)


@dataclass(frozen=True)
class Rule:
    operation: str
    operand: int

    def matches(self, value: int) -> bool:
        return cast("bool", OPS[self.operation](value, self.operand))


Input = list[AuntSue]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [AuntSue.from_input(line) for line in input_data]

    def find_aunt_sue_with_best_score(
        self, aunt_sues: list[AuntSue], rules: dict[str, Rule]
    ) -> int:
        scores = defaultdict[int, int](int)
        for aunt_sue, thing in itertools.product(aunt_sues, rules.keys()):
            value = aunt_sue.get_thing(thing)
            if value is not None and rules[thing].matches(value):
                scores[aunt_sue.nbr] += 1
        return max(scores.items(), key=operator.itemgetter(1))[0]

    def part_1(self, aunt_sues: Input) -> Output1:
        rules = {
            "children": Rule(EQ, 3),
            "cats": Rule(EQ, 7),
            "samoyeds": Rule(EQ, 2),
            "pomeranians": Rule(EQ, 3),
            "akitas": Rule(EQ, 0),
            "vizslas": Rule(EQ, 0),
            "goldfish": Rule(EQ, 5),
            "trees": Rule(EQ, 3),
            "cars": Rule(EQ, 2),
            "perfumes": Rule(EQ, 1),
        }
        return self.find_aunt_sue_with_best_score(aunt_sues, rules)

    def part_2(self, aunt_sues: Input) -> Output2:
        rules = {
            "children": Rule(EQ, 3),
            "cats": Rule("gt", 7),
            "samoyeds": Rule(EQ, 2),
            "pomeranians": Rule(LT, 3),
            "akitas": Rule(EQ, 0),
            "vizslas": Rule(EQ, 0),
            "goldfish": Rule(LT, 5),
            "trees": Rule(GT, 3),
            "cars": Rule(EQ, 2),
            "perfumes": Rule(EQ, 1),
        }
        return self.find_aunt_sue_with_best_score(aunt_sues, rules)

    def samples(self) -> None:
        pass


solution = Solution(2015, 16)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2015_16.java
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toMap;

import java.util.List;
import java.util.Map.Entry;
import java.util.Objects;
import java.util.function.Function;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import com.github.pareronia.aoc.IntegerSequence.Range;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2015_16 extends SolutionBase<List<AoC2015_16.AuntSue>, Integer, Integer> {
    
    private static final int[] VALUES = new int[] {
            /* Thing.CHILDREN */ 3,
            /* Thing.CATS */ 7,
            /* Thing.SAMOYEDS */ 2,
            /* Thing.POMERANIANS */ 3,
            /* Thing.AKITAS */ 0,
            /* Thing.VIZSLAS */ 0,
            /* Thing.GOLDFISH */ 5,
            /* Thing.TREES */ 3,
            /* Thing.CARS */ 2,
            /* Thing.PERFUMES */ 1 };
    
    private AoC2015_16(final boolean debug) {
        super(debug);
    }

    public static final AoC2015_16 create() {
        return new AoC2015_16(false);
    }

    public static final AoC2015_16 createDebug() {
        return new AoC2015_16(true);
    }
    
    private AuntSue findAuntSueWithBestScore(final List<AuntSue> auntSues, final Op[] ops) {
        return auntSues.stream()
            .collect(toMap(
                Function.identity(),
                sue -> IntStream.range(0, Thing.values().length)
                    .filter(thing -> ops[thing].matches(sue.getThings()[thing], VALUES[thing]))
                    .count()))
            .entrySet().stream()
                .sorted((e1, e2) -> Long.compare(e2.getValue(), e1.getValue()))
                .findFirst()
                .map(Entry::getKey)
                .orElseThrow();
    }

    @Override
    public List<AuntSue> parseInput(final List<String> inputs) {
        return inputs.stream().map(AuntSue::fromInput).collect(toList());
    }

    @Override
    public Integer solvePart1(final List<AuntSue> sues) {
        final Op[] ops = new Op[] {
                Op.EQ, Op.EQ, Op.EQ, Op.EQ, Op.EQ,
                Op.EQ, Op.EQ, Op.EQ, Op.EQ, Op.EQ };
        return findAuntSueWithBestScore(sues, ops).getNbr();
    }

    @Override
    public Integer solvePart2(final List<AuntSue> sues) {
        final Op[] ops = new Op[] {
                Op.EQ, Op.GT, Op.EQ, Op.LT, Op.EQ,
                Op.EQ, Op.LT, Op.GT, Op.EQ, Op.EQ };
        return findAuntSueWithBestScore(sues, ops).getNbr();
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2015_16.create().run();
    }
    
    private enum Op {
        EQ, LT, GT;
        
        public boolean matches(final Integer lhs, final int rhs) {
            if (this == Op.EQ) {
                return lhs != null && lhs == rhs;
            } else if (this == Op.LT) {
                return lhs != null && lhs < rhs;
            } else {
                return lhs != null && lhs > rhs;
            }
        }
    }

    enum Thing {
        CHILDREN("children"),
        CATS("cats"),
        SAMOYEDS("samoyeds"),
        POMERANIANS("pomeranians"),
        AKITAS("akitas"),
        VIZSLAS("vizslas"),
        GOLDFISH("goldfish"),
        TREES("trees"),
        CARS("cars"),
        PERFUMES("perfumes");
        
        private final String value;
        
        Thing(final String value) {
            this.value = value;
        }
        
        public static final Thing fromString(final String string) {
            return Stream.of(values())
                    .filter(v -> v.value.equals(string))
                    .findFirst().orElseThrow();
        }
    }

    static final class AuntSue {
        private final int nbr;
        private final Integer[] things;
        
        protected AuntSue(int nbr, Integer[] things) {
            this.nbr = nbr;
            this.things = things;
        }
        
        public static AuntSue fromInput(final String s) {
            final String[] splits = s.replaceAll("[,:]", "").split(" ");
            final Integer[] things = new Integer[Thing.values().length];
            Range.range(2, splits.length, 2).intStream()
                .forEach(i -> {
                   final int idx = Thing.fromString(splits[i]).ordinal();
                   things[idx] = Integer.valueOf(splits[i + 1]);
                });
            return new AuntSue(Integer.parseInt(splits[1]), things);
        }
        
        public int getNbr() {
            return nbr;
        }
        
        public Integer[] getThings() {
            return things;
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) {
                return true;
            }
            if (obj == null) {
                return false;
            }
            if (getClass() != obj.getClass()) {
                return false;
            }
            AuntSue other = (AuntSue) obj;
            return nbr == other.nbr;
        }

        @Override
        public int hashCode() {
            return Objects.hash(nbr);
        }
    }
}
